### PR TITLE
Improve forward compatibility between 2.x and 3.0

### PR DESCRIPTION
**Pull request by mdumoulin**
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | no
| Fixed issues | #4280

#### Summary

The current Doctrine\DBAL\Connection executeQuery(...) and executeCacheQuery() return a ResultStatement interface object. This PR try to improve forward compatibility between v2.x & v3.0 by adding some of missing features to Result returned by Connection :
- fetchNumeric()
- fetchAssociative()
- fetchOne()
- fetchAllNumeric()
- fetchAllAssociative()
- fetchAllKeyValue()
- fetchAllAssociativeIndexed()
- fetchFirstColumn()
- iterateNumeric()
- iterateAssociative()
- iterateKeyValue()
- iterateAssociativeIndexed()
- iterateColumn()
- rowCount()
- columnCount()
- free()

I used this version as reference : https://github.com/doctrine/dbal/blob/3.0.x/src/Result.php 

##### Static analysis error
Here under, methods which throw a static analysis error in phpstan but actually work. This is all methods defined in Abstraction\Result interface :

```php
/** @var Doctrine\DBAL\Connection $connection */
$connection->executeQuery('SELECT "something"')->fetchNumeric();
$connection->executeQuery('SELECT "something"')->fetchAssociative();
$connection->executeQuery('SELECT "something"')->fetchOne();
$connection->executeQuery('SELECT "something"')->fetchAllNumeric();
$connection->executeQuery('SELECT "something"')->fetchAllAssociative();
$connection->executeQuery('SELECT "something"')->fetchFirstColumn();
$connection->executeQuery('SELECT "something"')->iterateNumeric();
$connection->executeQuery('SELECT "something"')->iterateAssociative();
$connection->executeQuery('SELECT "something"')->iterateNumeric();
$connection->executeQuery('SELECT "something"')->iterateColumn();
```

The outpout pattern when running phpstan is the following :
```text
Call to an undefined method Doctrine\DBAL\Driver\ResultStatement::fetchOne().
```

#### Not available methods
Here under, methods which throw undefined method fatal & static analyse errors. This is all methods added in Abstraction\V3Result interface :

```php
$connection->executeQuery('SELECT "something"')->fetchAllKeyValue();
$connection->executeQuery('SELECT "something"')->fetchAllAssociativeIndexed();
$connection->executeQuery('SELECT "something"')->iterateKeyValue();
$connection->executeQuery('SELECT "something"')->iterateAssociativeIndexed();
```

The output pattern when trying to use one of those methods :
```text
Fatal error: Uncaught Error: Call to undefined method Doctrine\DBAL\Driver\PDO\Statement::fetchAllKeyValue() in /var/www/test.php on line 41

Error: Call to undefined method Doctrine\DBAL\Driver\PDO\Statement::fetchAllKeyValue() in /var/www/test.php on line 41

Call Stack:
    0.0006     415576   1. {main}() /var/www/test.php:0
```

#### Implementation key points

The strategy is to :
- Add a new interfaces with missing V3 Result features
- Make Connection return an object implementing ResultStatement & V3Result interfaces

### CODE REVIEW

File: lib/Doctrine/DBAL/Cache/ArrayStatement.php [hunk 1]
```diff
@@ -19,7 +19,7 @@
 /**
  * @deprecated
  */
-class ArrayStatement implements IteratorAggregate, ResultStatement, Result
+class ArrayStatement implements IteratorAggregate, ResultStatement
```
Comment (morozov) [inline lib/Doctrine/DBAL/Cache/ArrayStatement.php]: This is a breaking change. Please try using [Roave/BackwardCompatibilityCheck](https://github.com/Roave/BackwardCompatibilityCheck) (see https://github.com/doctrine/dbal/pull/4513 for example).

Comment (mdumoulin) [reply lib/Doctrine/DBAL/Cache/ArrayStatement.php]: Thanks for your quick return. It should be alright for BC break part now.

File: lib/Doctrine/DBAL/Cache/ArrayStatement.php [hunk 2]
```diff
@@ -19,7 +20,7 @@
 /**
  * @deprecated
  */
-class ArrayStatement implements IteratorAggregate, ResultStatement, Result
+class ArrayStatement implements IteratorAggregate, ResultStatement, V3Result
```
Comment (morozov) [inline lib/Doctrine/DBAL/Cache/ArrayStatement.php]: This is still a breaking change.

Comment (mdumoulin) [reply lib/Doctrine/DBAL/Cache/ArrayStatement.php]: Sorry, I don't understand why. This class still implements Result through V3Result. By the way, I used roave_bc_check tool and it did not throw BC break warning anymore. Can you explain it to me please ? 

Thanks again for your quick feedback.

Comment (morozov) [reply lib/Doctrine/DBAL/Cache/ArrayStatement.php]: Sorry, you're right. I didn't notice that `V3Result` extends `Result`. I'll try to review the changes more thoroughly in the next few days. In the meanwhile, please fix the Psalm failure and provide a couple of examples of the code that gets analyzed incorrectly w/o these changes.

File: lib/Doctrine/DBAL/Abstraction/V3Result.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace Doctrine\DBAL\Abstraction;
+
+use Doctrine\DBAL\Driver\Exception;
+use Traversable;
+
+/**
+ * @see https://github.com/doctrine/dbal/blob/3.0.x/src/Result.php
+ */
+interface V3Result extends Result
```
Comment (beberlei) [inline lib/Doctrine/DBAL/Abstraction/V3Result.php]: Couldn't this already be an interface called `Doctrine\DBAL\Result`? 

It could be documented as a docblock that it will turn from interface into class in 3.0.

File: lib/Doctrine/DBAL/Connection.php [hunk 4]
```diff
@@ -1235,7 +1194,7 @@ public function prepare($sql)
      * @param array<int, mixed>|array<string, mixed>                               $params Query parameters
      * @param array<int, int|string|Type|null>|array<string, int|string|Type|null> $types  Parameter types
      *
-     * @return ResultStatement The executed statement.
+     * @return ResultStatement&V3Result The executed statement.
```
Comment (beberlei) [inline lib/Doctrine/DBAL/Connection.php]: I would rather as mentioned above this being `Doctrine\DBAL\Result`, because that is the type it will become in 3.0

**Review by beberlei — COMMENTED** [review]
This is very important, however I think we should be bolder and move the `V3Result` to be `Doctrine\DBAL\Result` otherwise the API will change again in 3.0 and you cannot rely on typing the result tpye, example:

```php
use Doctrine\DBAL\Result;

public function performQuery()
{
     $result = $this->connection->executeQuery('SELECT * FROM orders');
     return $this->convertQueryResult($result);
}

private function convertQueryResult(Result $result)
{
}
```

File: lib/Doctrine/DBAL/Abstraction/V3Result.php (hunk 3 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Abstraction/V3Result.php]: This should work. Otherwise, I'd rather use a more specific term like `ForwardCompatibility` and the corresponding namespace (see #4019) to simplify the cleanup.

File: lib/Doctrine/DBAL/Connection.php [hunk 5]
```diff
@@ -1332,6 +1291,18 @@ public function executeCacheQuery($sql, $params, $types, QueryCacheProfile $qcp)
 
         $stmt->setFetchMode($this->defaultFetchMode);
 
+        return $this->createV3CompatStatement($stmt);
+    }
+
+    /**
+     * @return ResultStatement&V3Result
+     */
+    private function createV3CompatStatement(ResultStatement $stmt)
```
Comment (morozov) [inline lib/Doctrine/DBAL/Connection.php]: Please use "forward compatibility" instead of "v3". Also, the method doesn't unconditionally create a statement as the name implies. Would something like `ensureStatementForwardCompatibility()` make sense.

File: lib/Doctrine/DBAL/Connection.php [hunk 6]
```diff
@@ -1332,6 +1291,18 @@ public function executeCacheQuery($sql, $params, $types, QueryCacheProfile $qcp)
...
+    {
+        if (! $stmt instanceof V3Result) {
```
Comment (morozov) [inline lib/Doctrine/DBAL/Connection.php]: Negation can be avoided to improve readability:
```php
if ($stmt instanceof V3Result) {
    return $stmt;
}

return new V3CompatStatement($stmt);
```

File: lib/Doctrine/DBAL/Driver/V3CompatStatement.php [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+namespace Doctrine\DBAL\Driver;
+
+use Doctrine\DBAL\Abstraction\V3Result;
+use Doctrine\DBAL\Exception;
+use IteratorAggregate;
+use PDO;
+use Traversable;
+
+class V3CompatStatement implements IteratorAggregate, ResultStatement, V3Result
```
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/V3CompatStatement.php]: Please move to a separate `ForwardCompatibility` namespace.

File: lib/Doctrine/DBAL/Driver/V3CompatStatement.php [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    {
+        return $this->stmt->setFetchMode($fetchMode, $arg2, $arg3);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetch($fetchMode = null, $cursorOrientation = PDO::FETCH_ORI_NEXT, $cursorOffset = 0)
+    {
+        return $this->stmt->fetch($fetchMode, $cursorOrientation, $cursorOffset);
```
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/V3CompatStatement.php]: I don't know if it's applicable to this case but in some cases, just passing the arguments down the line doesn't work because the underlying driver relies on the number of arguments passed, not their values: https://github.com/doctrine/dbal/blob/84553e361d7ab15b16c45e1a08572d07e8a992e0/lib/Doctrine/DBAL/Driver/PDOStatement.php#L127-L136

The fact that the build is passing means that most likely this case isn't tested.

File: lib/Doctrine/DBAL/Driver/V3CompatStatement.php [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    {
+        return FetchUtils::iterateColumn($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function rowCount()
+    {
+        throw Exception::notSupported('rowCount');
```
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/V3CompatStatement.php]: Doesn't it make the implementation incompatible with the old one? It is possible to count rows on the statement in 2.x.

File: lib/Doctrine/DBAL/Connection.php [hunk 10]
```diff
@@ -618,11 +619,7 @@ public function fetchAssociative(string $query, array $params = [], array $types
         try {
             $stmt = $this->executeQuery($query, $params, $types);
 
-            if ($stmt instanceof Result) {
```
Comment (morozov) [inline lib/Doctrine/DBAL/Connection.php:621]: What if `executeQuery()` is overridden in a custom class and doesn't implement the newly introduced wrapping?

File: lib/Doctrine/DBAL/Driver/FetchUtils.php [hunk 11]
```diff
@@ -57,6 +62,44 @@ public static function fetchAllAssociative(Result $result): array
...
     }
 
+    /**
+     * Returns an array containing the values of the first column of the result.
+     *
+     * @return array<mixed,mixed>
+     *
+     * @throws Exception
+     */
+    public static function fetchAllAssociativeIndexed(Result $result): array
```
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/FetchUtils.php]: I'd rather have these methods implemented in a separate class under the `ForwardCompatibility` namespace. This way, it will be clear that it won't need to be merged up. Also, it reduces the risk of accidentally introducing an API that isn't supported in `3.x`.

File: lib/Doctrine/DBAL/Driver/IBMDB2/DB2Statement.php [hunk 12]
```diff
@@ -427,6 +428,22 @@ public function fetchAllAssociative(): array
         return FetchUtils::fetchAllAssociative($this);
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function fetchAllKeyValue(): array
```
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/IBMDB2/DB2Statement.php]: Drivers do not implement this method in `3.x`. Why does it need to be implemented in `2.x`?

File: lib/Doctrine/DBAL/Abstraction/V3Result.php (hunk 3 above)
Comment (mdumoulin) [reply lib/Doctrine/DBAL/Abstraction/V3Result.php]: Seems a good idea to name it after the name used in 3.0.

File: lib/Doctrine/DBAL/Connection.php (hunk 5 above)
Comment (mdumoulin) [reply lib/Doctrine/DBAL/Connection.php]: Done

File: lib/Doctrine/DBAL/Connection.php (hunk 6 above)
Comment (mdumoulin) [reply lib/Doctrine/DBAL/Connection.php]: Done

File: lib/Doctrine/DBAL/Driver/V3CompatStatement.php (hunk 7 above)
Comment (mdumoulin) [reply lib/Doctrine/DBAL/Driver/V3CompatStatement.php]: Done

File: lib/Doctrine/DBAL/Driver/IBMDB2/DB2Statement.php (hunk 12 above)
Comment (mdumoulin) [reply lib/Doctrine/DBAL/Driver/IBMDB2/DB2Statement.php]: It is not available in the driver implementation, but feature is ensured by the Result class : https://github.com/doctrine/dbal/blob/3.0.x/src/Result.php#L120-L131

It seems to me we have to ensure that user can write the following :
```php
$result = $connection->executeQuery(...)->fetchAllKeyValue();
```

File: lib/Doctrine/DBAL/Driver/FetchUtils.php (hunk 11 above)
Comment (mdumoulin) [reply lib/Doctrine/DBAL/Driver/FetchUtils.php]: Added methods are used exactly in the same way that previously existing ones : fetchOne(), fetchAllNumeric(), fetchAllAssociative() ... etc. Wouldn't it be weird to split them in 2 separate classes ?

File: lib/Doctrine/DBAL/Connection.php (hunk 10 above)
Comment (mdumoulin) [reply lib/Doctrine/DBAL/Connection.php:621]: You are right, it will break this kind of extension. I propose to use the ensureForwardCompatibleStatement() that should do the work.

File: lib/Doctrine/DBAL/Driver/V3CompatStatement.php (hunk 9 above)
Comment (mdumoulin) [reply lib/Doctrine/DBAL/Driver/V3CompatStatement.php]: The class decorate the ResultStatement which does not contain this method but some subclasses implements it. Maybe a simple method_exists() check can do the job ?

File: lib/Doctrine/DBAL/Driver/V3CompatStatement.php (hunk 8 above)
Comment (mdumoulin) [reply lib/Doctrine/DBAL/Driver/V3CompatStatement.php]: Not sure to fully understand in which case it could cause an issue. Anyway, all drivers statements of the library have been migrated on forward compatible Result interface, so they will not be wrapped in the compatibility statement object.

https://github.com/doctrine/dbal/blob/c4e4ce67bde3d407db27b157306373d52a6b0bec/lib/Doctrine/DBAL/Connection.php#L1308-L1315

File: lib/Doctrine/DBAL/Driver/FetchUtils.php (hunk 11 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Driver/FetchUtils.php]: The point is that drivers must implement only a subset of the methods while the ones you're adding are implemented at the wrapper level. You're adding wrapper-specific methods to the driver level.

File: lib/Doctrine/DBAL/Driver/IBMDB2/DB2Statement.php (hunk 12 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Driver/IBMDB2/DB2Statement.php]: It creates a BC break in 3.0.0 because these methods won't exist there. The driver layer must implement only a subset of `fetch*()` methods and none of the `iterate*()` ones.

Comment (mdumoulin) [reply lib/Doctrine/DBAL/Driver/IBMDB2/DB2Statement.php]: Ok, I did not see that but it make perfect senses. I will remove the support of Doctrine\PDO\Result interface by Drivers and use an object to wrap the Driver\Result as it is done in 3.0.

File: lib/Doctrine/DBAL/Cache/ArrayStatement.php [hunk 13]
```diff
@@ -4,12 +4,13 @@
 
 use ArrayIterator;
 use Doctrine\DBAL\Driver\FetchUtils;
-use Doctrine\DBAL\Driver\Result;
 use Doctrine\DBAL\Driver\ResultStatement;
 use Doctrine\DBAL\FetchMode;
+use Doctrine\DBAL\Result;
```
Comment (morozov) [inline lib/Doctrine/DBAL/Cache/ArrayStatement.php]: An array statement must implement the driver-level API, not the wrapper level one.

File: lib/Doctrine/DBAL/Cache/ArrayStatement.php [hunk 14]
```diff
@@ -215,6 +216,22 @@ public function fetchAllAssociative(): array
...
+     */
+    public function fetchAllKeyValue(): array
+    {
+        return FetchUtils::fetchAllKeyValue($this);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function fetchAllAssociativeIndexed(): array
```
Comment (morozov) [inline lib/Doctrine/DBAL/Cache/ArrayStatement.php]: This method doesn't belong here.

File: lib/Doctrine/DBAL/Cache/ResultCacheStatement.php [hunk 15]
```diff
@@ -35,7 +37,7 @@
  *
  * @deprecated
  */
-class ResultCacheStatement implements IteratorAggregate, ResultStatement, Result
+class ResultCacheStatement implements IteratorAggregate, ResultStatement, ForwardCompatibilityResult
```
Comment (morozov) [inline lib/Doctrine/DBAL/Cache/ResultCacheStatement.php]: It shouldn't implement the wrapper-level result.

**Comment by mdumoulin** [comment]
@morozov Drivers statements no longer implements wrapper level Result. It break some unit tests because of ResultStatement wrapper. The following code will no longer work for example.

```php
$statement = $connection->executeQuery('SELECT "something"');

if ($statement instanceof ArrayStatement) {
    // True in 2.12.x version, false in this branch
}
```

I can't see any generic solution for this kind of edge case. Maybe we should let the user enable the Statement wrapper on purpose. We will avoid the risk of regression and allow to resolve deprecation on 2.x. It can be implemented by adding a ForwardCompatibility\Connection wrapper for example.

**Comment by morozov** [comment]
> Maybe we should let the user enable the Statement wrapper on purpose. We will avoid the risk of regression and allow to resolve deprecation on 2.x. It can be implemented by adding a ForwardCompatibility\Connection wrapper for example.

If it's an opt-in layer, then there will be no improvement for static analysis, because the analyzers won't be able to detect the opt-in. As for the examples that currently don't work, e.g.:
```php
$connection->executeQuery('SELECT "something"')->fetchNumeric();
```
There's already an alternative:
```php
$connection->fetchNumeric('SELECT "something"');
```

**Comment by beberlei** [comment]
Yes this edge case break is ok imho, the api is returning a statement, expecting a specific statement is not part of the interface

File: tests/Doctrine/Tests/DBAL/Functional/ResultCacheTest.php [hunk 16]
```diff
@@ -238,25 +238,25 @@ public function testFetchingAllRowsSavesCache(callable $fetchAll): void
     public static function fetchAllProvider(): iterable
     {
         yield 'fetchAll' => [
-            static function (ResultCacheStatement $statement): void {
+            static function (ForwardCompatibility\Result $statement): void {
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/ResultCacheTest.php]: The expected `ResultCacheStatement` type was unnecessarily specific. Let's loosen it to `Driver\ResultStatement`. If it had been implemented this way, it wouldn't require a change now.

**Review by morozov — COMMENTED** [review]
Looks good.

File: tests/Doctrine/Tests/DBAL/Functional/ResultCacheTest.php [hunk 17]
```diff
@@ -238,25 +238,25 @@ public function testFetchingAllRowsSavesCache(callable $fetchAll): void
...
         yield 'fetchAllNumeric' => [
-            static function (ResultCacheStatement $statement): void {
+            static function (ForwardCompatibility\Result $statement): void {
                 $statement->fetchAllNumeric();
             },
         ];
 
         yield 'fetchFirstColumn' => [
-            static function (ResultCacheStatement $result): void {
+            static function (ForwardCompatibility\Result $result): void {
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/ResultCacheTest.php]: https://github.com/doctrine/dbal/pull/4529#discussion_r593785306 applies to all the above cases.

Comment (mdumoulin) [reply tests/Doctrine/Tests/DBAL/Functional/ResultCacheTest.php]: Not sure to understand here. ResultStament interface does not contain the fetchAllAssociative(), fetchAllNumeric() &  fetchFirstColumn() methods. I can use the Doctrine\DBAL\Result interface, if we prefer not relying on concrete object.

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/ResultCacheTest.php]: Sorry, it should have been `Driver\Result`.

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/ResultCacheTest.php]: Alright, so the expected type was not unnecessarily specific: it was the type that implemented both of the old and new APIs. It still implements them but now the `fetch*()` methods return another class that implements both interfaces as well. Looks good 👍

**Comment by morozov** [comment]
@mdumoulin It looks good. I'd like all the changes to be squashed in one commit. Please do or otherwise, I'll do it later before the merge.

**Comment by mdumoulin** [comment]
@morozov Great ! I just squashed all my commits in a single one.

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thanks, @mdumoulin!

**Comment by mdumoulin** [comment]
@morozov @beberlei Thanks again for your quick feedback & review.
